src/components/shared/WithSize.js

```diff
@@ -1,44 +1,59 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
 
-import React, { PureComponent } from 'react';
+import * as React from 'react';
 import { findDOMNode } from 'react-dom';
+import type { CssPixels } from '../../types/units';
+
+export type SizeProps = {|
+  width: CssPixels,
+  height: CssPixels,
+|};
 
 /**
  * Wraps a React component and makes 'width' and 'height' available in the
  * wrapped component's props. These props start out at zero and are updated to
  * the component's DOM node's getBoundingClientRect().width/.height after the
  * component has been mounted. They also get updated when the window is
  * resized.
+ *
  * Note that the props are *not* updated if the size of the element changes
  * for reasons other than a window resize.
- * @param  {class} Wrapped The class that gets wrapped.
- * @return {class}         The resulting Component class.
  */
-export const withSize = Wrapped =>
-  class WithSizeWrapper extends PureComponent {
-    constructor(props) {
-      super(props);
-      this.state = { width: 0, height: 0 };
-      this._observeSize = this._observeSize.bind(this);
-    }
+export function withSize<WrappedProps: Object>(
+  Wrapped: React.ComponentType<{ ...WrappedProps, ...SizeProps }>
+): React.ComponentType<{ ...WrappedProps, ...SizeProps }> {
```

Comment:
this looks weird: the returned component shouldn't have SizeProps.

I just tried: if you replace the return value by `React.ComponentType<WrappedProps>` Flow still passes. What is weird to me is why it was passing before.

Reply:
I tried things locally and when I typed what I think should be typed, flow is outputting a lot of errors.
I really believe the exact types are making things too complicated and maybe impossible to type at the moment, but I don't want to hold the patch for this.

At the moment I believe this typing doesn't check much. For example try to add a property to TimeSelectionScrubber in ProfileViewerHeader => Flow doesn't error.
Same if you _remove_ a property (like `zeroAt`).

Same in WithSize: you can add a random property to `Wrapped` and Flow doesn't care. Same if you remove `{...this.state}`.

I think we can keep it like this in this patch to move forward but we'll need to come back at this. I think that HOC don't work well with exact props, too bad but that's how it is.  And then I think we need to follow closely what's in https://flow.org/en/docs/react/hoc/.

Note: that's why I was suggesting something more complicated in my initial review, by typing both the wrapped component and the returned component, so that we can still get exact objects, but I don't know if Flow would allow it. So maybe not do it for now.

TLDR: keep it like this for this patch as this is complicated but please file a bug.